dagger/build_bins.go

```diff
@@ -0,0 +1,325 @@
+package main
+
+import (
+	"context"
+	"dagger/embedded-cluster/internal/dagger"
+	"fmt"
+	"strings"
+)
+
+// BuildBin builds the web UI and binary (without embedding the release).
+//
+// This is equivalent to ci-build-bin.sh and builds:
+// - Web UI (React/TypeScript frontend)
+// - Go binary with embedded web assets
+// - Metadata JSON
+//
+// Example:
+//
+//	dagger call with-one-password --service-account=env:OP_SERVICE_ACCOUNT_TOKEN \
+//	  with-build-metadata --dir ./output \
+//	  build-bin \
+//	  build-metadata to-dir export --path=./output
+func (m *EmbeddedCluster) BuildBin(
+	ctx context.Context,
+	// Source directory to use for the build.
+	// +defaultPath="/"
+	// +optional
+	src *dagger.Directory,
+	// S3 bucket for artifact URLs
+	// +default="dev-embedded-cluster-bin"
+	s3Bucket string,
+	// GitHub token
+	// +optional
+	githubToken *dagger.Secret,
+) (*EmbeddedCluster, error) {
+	if m.BuildMetadata == nil {
+		return nil, fmt.Errorf("build metadata is required")
+	}
+	if m.BuildMetadata.Version == "" {
+		return nil, fmt.Errorf("version is required")
+	}
+	if m.BuildMetadata.K0SMinorVersion == "" {
+		return nil, fmt.Errorf("k0s minor version is required")
+	}
+	if m.BuildMetadata.Arch == "" {
+		return nil, fmt.Errorf("arch is required")
+	}
+
+	// Build web UI
+	webBuild := m.BuildWeb(ctx, src)
+
+	// Build binary
+	buildResult, err := m.buildBinary(
+		ctx,
+		src,
+		webBuild,
+		s3Bucket,
+		githubToken,
+		s3Bucket != StagingS3Bucket,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Update metadata with build artifacts
+	m.BuildMetadata.BuildDir = buildResult
+
+	return m, nil
+}
+
+// buildBinary builds the embedded-cluster binary (ports ci-build-bin.sh logic)
+func (m *EmbeddedCluster) buildBinary(
+	ctx context.Context,
+	src *dagger.Directory,
+	webBuild *dagger.Directory,
+	s3Bucket string,
+	githubToken *dagger.Secret,
+	usesDevBucket bool,
+) (*dagger.Directory, error) {
+	// Use cached build environment
+	builder := m.buildEnv(src)
+
+	// Set build environment variables
+	builder = m.BuildMetadata.withEnvVariables(builder).
+		WithEnvVariable("IMAGES_REGISTRY_SERVER", "ttl.sh")
```

Comment:
hmm this (ttl.sh) might not work in the production ci pipeline as we publish the image to dockerhub currently. same for the LAM image